diff --git a/miners/text/template_server.py b/miners/text/template_server.py
index e176edd..fe53335 100644
--- a/miners/text/template_server.py
+++ b/miners/text/template_server.py
@@ -27,6 +27,7 @@ import torch
 import time
 import wandb
 import datetime
+from qqdm import qqdm
 from transformers import BertModel, BertConfig
 
 def config ():
@@ -122,14 +123,15 @@ def main( config ):
         while True:
             metagraph.sync().save()
             uid = metagraph.hotkeys.index( wallet.hotkey.ss58_address )
-            wandb.log( 
-                {
-                    'Stake': metagraph.S[ uid ].item(),
-                    'Rank': metagraph.R[ uid ].item(),
-                    'Incentive': metagraph.I[ uid ].item(),
-                    'Axon QPS': axon.stats.qps.value
-                } 
-            ) 
+            wandb_data = {
+                'Stake': metagraph.S[ uid ].item(),
+                'Rank': metagraph.R[ uid ].item(),
+                'Incentive': metagraph.I[ uid ].item(),
+                'Axon QPS': axon.stats.qps.value
+            } 
+            for uid_i, val in enumerate(metagraph.W[:,uid].tolist()):
+                wandb_data[ 'w_\{{},{}\}'.format(uid_i, uid) ] = val
+            wandb.log( wandb_data )
             time.sleep( 30 * bittensor.__blocktime__ )
 
 if __name__ == "__main__":
